pkg/template/template.go

```diff
@@ -0,0 +1,162 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package template
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"strings"
+	tpl "text/template"
+
+	"github.com/youmark/pkcs8"
+	"golang.org/x/crypto/pkcs12"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var tplFuncs = tpl.FuncMap{
+	"pkcs12key":      pkcs12key,
+	"pkcs12keyPass":  pkcs12keyPass,
+	"pkcs12cert":     pkcs12cert,
+	"pkcs12certPass": pkcs12certPass,
+
+	"pemPrivateKey":  pemPrivateKey,
+	"pemCertificate": pemCertificate,
+	"base64decode":   base64decode,
+	"base64encode":   base64encode,
+	"fromJSON":       fromJSON,
+	"toJSON":         toJSON,
+
+	"toString": toString,
+	"toBytes":  toBytes,
+	"upper":    strings.ToUpper,
+	"lower":    strings.ToLower,
+}
+
+var log = ctrl.Log.WithName("template")
+
+// Execute uses an best-effort approach to render the secret data as template.
+func Execute(secret *corev1.Secret, data map[string][]byte) error {
+	for k, v := range secret.Data {
+		t, err := tpl.New(k).
+			Funcs(tplFuncs).
```

Comment:
Do we want to use https://github.com/Masterminds/sprig template functions? This would match the library and functions that Helm has. This library includes `upper`, `lower` and a few of the other functions above.

Reply:
I missed the above conversation, I would suggest that we further evaluate sprig as I think it will be useful to have matching functionality of Helm or other projects in this space. Looks like Helm just removes the unsafe `env` functions as was mentioned.

```golang
f := sprig.TxtFuncMap()
delete(f, "env")
delete(f, "expandenv")
```

I'm good with merging this in and continuing the sprig conversation in another issue :+1: